pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,201 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CalicoNodeStatusList is a list of CalicoNodeStatus resources.
+type CalicoNodeStatusList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []CalicoNodeStatus `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type CalicoNodeStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   CalicoNodeStatusSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status CalicoNodeStatusStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// CalicoNodeStatusSpec contains the specification for a CalicoNodeStatus resource.
+type CalicoNodeStatusSpec struct {
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
+	Classes []NodeStatusClassType `json:"classes,omitempty"`
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
+}
+
+// CalicoNodeStatusStatus defines the observed state of CalicoNodeStatus.
+// No validation needed for status since it is updated by Calico.
+type CalicoNodeStatusStatus struct {
+	// LastUpdated is a timestamp representing the server time when CalicoNodeStatus object
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	// +nullable
+	LastUpdated metav1.Time `json:"lastUpdated,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
+
+	// Agent holds agent status on the node.
+	Agent CalicoNodeAgentStatus `json:"agent,omitempty"`
+
+	// BGP holds node BGP status.
+	BGP CalicoNodeBGPStatus `json:"bgp,omitempty"`
+
+	// Route represents routes on the node.
+	Route CalicoNodeRouteStatus `json:"route,omitempty"`
+}
+
+// CalicoNodeAgentStatus defines the observed state of agent status on the node.
+type CalicoNodeAgentStatus struct {
+	// Bird4 represents the latest observed status of bird4.
+	Bird4 CalicoNodeBirdStatus `json:"bird4,omitempty"`
+
+	// Bird6 represents the latest observed status of bird6.
+	Bird6 CalicoNodeBirdStatus `json:"bird6,omitempty"`
+}
+
+// CalicoNodeBGPStatus defines the observed state of BGP status on the node.
+type CalicoNodeBGPStatus struct {
+	// The total number of established bgp sessions.
+	NumEstablished int `json:"numEstablished,omitempty"`
+
+	// The total number of non-established bgp sessions.
+	NumNotEstablished int `json:"numNotEstablished,omitempty"`
```

Comment:
I think I'd probably want a separate `NumNotEstablished` field for v4 and v6.  Then, in a typical system where I'm only interested in v4, I can implement a very simple sanity check as `v4.NumNotEstablished == 0`.

Reply:
Whereas if I see NumEstab 2 NumNot 2, it probably means missing IPv6 support, but I can't immediately be sure about that.

Reply:
Yes. Make sense.